game-headed/src/main/java/org/triplea/game/client/ui/javafx/screens/RoleSelection.java

```diff
@@ -0,0 +1,192 @@
+                .orElseThrow(
+                    () -> new IllegalStateException("Missing GameData when calling screen"));
+    final List<String> availablePlayers = List.of(PlayerType.playerTypes());
+    for (final GamePlayer gamePlayer : gameData.getPlayerList()) {
+      setupPlayerControl(availablePlayers, gamePlayer);
+    }
+  }
+
+  private void setupPlayerControl(
+      final List<String> availablePlayers, final GamePlayer gamePlayer) {
+    final var name = new Label(gamePlayer.getName());
+    final ComboBox<String> controllingPlayer =
+        new ComboBox<>(FXCollections.observableArrayList(availablePlayers));
+    controllingPlayer.getSelectionModel().select(0);
+    if (gamePlayer.getCanBeDisabled()) {
+      controllingPlayer.getItems().add(DISABLE_TEXT);
+    }
+    final Button faction = newFactionButton(gamePlayer, controllingPlayer);
+    roleForPlayers.put(gamePlayer, controllingPlayer);
+    final var income = new Spinner<Integer>(0, 100, 100);
+    income.setDisable(true);
+    incomeForPlayers.put(gamePlayer, income);
+    final var pus = new Spinner<Integer>(0, 100, 0);
+    pus.setDisable(true);
+    pusForPlayers.put(gamePlayer, pus);
+    factionGrid.addRow(factionGrid.getRowCount(), name, controllingPlayer, faction, income, pus);
+  }
+
+  private Button newFactionButton(
+      final GamePlayer gamePlayer, final ComboBox<String> controllingPlayer) {
+    final Collection<String> playerAlliances =
+        gameData.getAllianceTracker().getAlliancesPlayerIsIn(gamePlayer);
+    final var faction = new Button(playerAlliances.toString());
+    faction.setOnAction(
+        e -> {
+          final int targetIndex = controllingPlayer.getSelectionModel().getSelectedIndex();
+          playerAlliances.stream()
+              .map(gameData.getAllianceTracker()::getPlayersInAlliance)
+              .flatMap(Collection::stream)
+              .map(roleForPlayers::get)
+              .filter(stringComboBox -> stringComboBox.getItems().size() > targetIndex)
+              .map(ComboBox::getSelectionModel)
+              .forEach(selectionModel -> selectionModel.select(targetIndex));
+        });
+    return faction;
+  }
+
+  @Override
+  public Map<String, Class<?>> getValidTypes() {
+    return Map.of(SELECTED_MAP_KEY, GameData.class);
+  }
+
+  @FXML
+  void setAllTo() {
+    factionGrid.getChildren().stream()
+        .filter(ComboBox.class::isInstance)
+        .map(node -> (ComboBox<?>) node)
+        .forEach(
+            comboBox ->
+                comboBox
+                    .getSelectionModel()
+                    .select(allSelectorCheckbox.getSelectionModel().getSelectedIndex()));
+  }
+
+  @FXML
+  void cancelMapSelection() {
+    screenController.switchScreen(FxmlManager.MAP_SELECTION);
+  }
+
+  @FXML
+  void startGame() {
+    if (resourceModifierCheckbox.isSelected()) {
+      setupPuIncome();
+    }
+    final List<Entry<String, String>> flatMapping =
```

Comment:
It looks like we have a decent amount of logic in the FXML class here. I was looking at the function reference "DISABLE_TEXT::equals", thinking that was logic in the XML class, but going up further, the `startGame` method itself has logic. This looks like a place where we are mixing display with back-end logic. 

I think we could have a really good fit for view-model, model-controller pattern here. If we had one class to represent the FXML components, I think we could decouple quite nicely and also make the logic test-able. 

EG:
```
roleSelectionViewModel.setResourceModifierSelected(resourceModifierCheckbox.isSelected());
roleSelectionViewModel.setPlayerRoles(roleForPlayers);
roleSelectionViewModel.startGame();
```

Have you considered using a view-model, any thoughts?

Reply:
I feel like the line between what should be UI-only and what is logic is really blurry here.
Your suggestion sounds intruiging, but I'm not quite sure how that would look like for the cases where the view has to query the model for information, for example when pressing on any alliance button next to the control-type selection (AI/Human/Disabled).
Also I'm not quite happy with the disable text thingy either, I simply haven't found a good way to store the intermediate selection state and synchronize it with the UI, so I just try to convert the UI back to an "internal" repreentation we can work with.
Maybe that's just me and this isn't a problem at all, but if you can show me a good pattern that works both ways (from view to model and vice-versa) then I'd happily apply it everywhere.

Reply:
Typically model interactions are mediated by the view model (afaik). Let's keep thinking about this, both read up on view-model pattern a bit. view-model is also agnostic about UI layer, which makes it really great for decoupling swing and potentially allowing more migration of UI. In essence it also allows for UI code to be far more pure, and hence better decoupling. I think there are some examples in the forum poster code where I'll try a view-model, it should give an example of the pattern as well, so let's keep this on our radar.